x/reward/keeper/query_server.go

```diff
@@ -0,0 +1,244 @@
+		if errFromParsingKey != nil {
+			return false, err
+		}
+		result, errFromGetRewardAccount := k.GetRewardAccountState(sdkCtx, lookupVal.RewardID, lookupVal.ClaimID, lookupVal.Addr.String())
+		// think ignoring the error maybe ok here since it's just another lookup
+		if errFromGetRewardAccount != nil {
+			return false, nil
+		}
+		if result.GetSharesEarned() == 0 || (request.ClaimStatus != result.ClaimStatus && request.ClaimStatus != types.RewardAccountState_CLAIM_STATUS_UNSPECIFIED) {
+			return false, nil
+		}
+
+		if accumulate {
+			states = append(states, result)
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrIterateAllRewardAccountStates, err.Error())
+	}
+
+	rewardAccountResponses := k.convertRewardAccountStateToRewardAccountResponse(sdkCtx, states)
+	rewardAccountByAddressResponse := types.QueryRewardDistributionsByAddressResponse{
+		Address:            request.Address,
+		RewardAccountState: rewardAccountResponses,
+		Pagination:         pageRes,
+	}
+
+	return &rewardAccountByAddressResponse, nil
+}
+
+func (k Keeper) convertRewardAccountStateToRewardAccountResponse(ctx sdk.Context, states []types.RewardAccountState) []types.RewardAccountResponse {
+	rewardAccountResponse := make([]types.RewardAccountResponse, 0)
+	for _, state := range states {
+		rewardProgram, err := k.GetRewardProgram(ctx, state.GetRewardProgramId())
+		if err != nil {
+			continue
+		}
+		distribution, err := k.GetClaimPeriodRewardDistribution(ctx, state.ClaimPeriodId, state.RewardProgramId)
+		if err != nil {
+			continue
+		}
+
+		participantReward := k.CalculateParticipantReward(ctx, int64(state.GetSharesEarned()), distribution.GetTotalShares(), distribution.GetRewardsPool(), rewardProgram.MaxRewardByAddress)
+		accountResponse := types.RewardAccountResponse{
+			RewardProgramId:  state.RewardProgramId,
+			TotalRewardClaim: participantReward,
+			ClaimStatus:      state.ClaimStatus,
+			ClaimId:          state.ClaimPeriodId,
+		}
+		rewardAccountResponse = append(rewardAccountResponse, accountResponse)
+	}
+
+	return rewardAccountResponse
+}
+
+// hasPageRequest is just for use with the getPageRequest func below.
+type hasPageRequest interface {
+	GetPagination() *query.PageRequest
+}
+
+// Gets the query.PageRequest from the provided request if there is one.
+// Also sets the default limit if it's not already set yet.
+func getPageRequest(req hasPageRequest) *query.PageRequest {
+	var pageRequest *query.PageRequest
+	if req.GetPagination() != nil {
+		pageRequest = req.GetPagination()
+		// enforce max/min page limit
+		enforceMaxMinPageLimit(pageRequest)
+		return pageRequest
+	}
+	if pageRequest == nil {
+		pageRequest = &query.PageRequest{}
+	}
+	enforceMaxMinPageLimit(pageRequest)
+	return pageRequest
+}
+
+func enforceMaxMinPageLimit(pageRequest *query.PageRequest) {
+	if pageRequest.Limit == 0 || pageRequest.Limit > defaultPerPageLimit {
+		pageRequest.Limit = defaultPerPageLimit
+	}
+}
```

Comment:
We modified an example of the metadata modules pagination pattern here.  Yes, it was giving us a few problems.  I just stripped everything out and followed how the sdk does things instead of adding this custom query methods.  The sdk will set the default max limit to `100` if it isn't set.

Reply:
But this sets a **MAX** of 100 on the limit, which isn't very user friendly. E.g. If I do a query with `--limit 500`, this will then only return 100 entries (even if there are more than that).

With the SDK's logic, providing `Limit = 0` (or not providing a limit argument) is the same as providing `Limit = 100` and `CountTotal = true`. By defaulting it to `100` here (before it gets to the SDK logic), you lose the `CountTotal = true` behavior. If that's what's intended here, there should be a comment about it with a bit about why.

But even if the change to `CountTotal` is intended, there shouldn't be a max on the limit.

Reply:
Sorry, I wasn't clear.  I saw that as a problem as well.  I removed all that code and just let the sdk handle the pagination.  We are no longer calling the problematic method (`enforceMaxMinPageLimit`)  that set the **MAX** as 100.  Also, the sdk will set the limit to 100 if the `Limit = 0`.  Therefore, checking if the limit is set to 0 ourself and setting it to 100 is redundant.